libbeat/licenser/manager.go

```diff
@@ -0,0 +1,279 @@
+	"sync"
+	"time"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+// OSSLicense default license to use.
+var OSSLicense = &License{UUID: uuid.NewV4(), Type: OSS, Mode: OSS, Status: Active}
+
+// Watcher allows a type to receive a new event when a new license is received.
+type Watcher interface {
+	OnNewLicense(license License)
+	OnManagerStopped()
+}
+
+// Fetcher interface implements the mechanism to retrieve a License. Currently we only
+// support license coming from the '/_xpack' rest api.
+type Fetcher interface {
+	Fetch() (*License, error)
+}
+
+// Errors returned by the manager.
+var (
+	ErrWatcherAlreadyExist = errors.New("watcher already exist")
+	ErrWatcherDoesntExist  = errors.New("watcher doesn't exist")
+
+	ErrManagerStopped = errors.New("license manager is stopped")
+	ErrNoLicenseFound = errors.New("no license found")
+)
+
+// Backoff values when the remote cluster is not responding.
+var (
+	maxBackoff  = time.Duration(60)
+	initBackoff = time.Duration(5)
+)
+
+// Manager keeps tracks of license management, it uses a fetcher usually the ElasticFetcher to
+// retrieve a licence from a specific cluster.
+//
+// Starting the manager will starts a go routine to periodically query the license fetcher.
+// if an error occur on the fetcher we will retries until we successfully
+// receive a new license. During that period we start a grace counter, we assume the license is
+// still valid during the grace period, when this period expire we will keep retrying but the previous
+// license will be invalidated and we will fallback to the OSS license.
+//
+// Retrieving the current license:
+// - Call the `Get()` on the manager instance.
+// - Or register a `Watcher` with the manager to receive the new license and acts on it, you will
+// also receive an event when the Manager is stopped.
+//
+//
+// Notes:
+// - When the manager is started no license is set by default.
+// - When a license is invalidated, we fallback to the OSS License and the watchers get notified.
+// - Adding a watcher will automatically send the current license to the newly added watcher if
+//   available.
+type Manager struct {
+	done chan struct{}
+	sync.RWMutex
+	fetcher     Fetcher
+	duration    time.Duration
+	gracePeriod time.Duration
+	license     *License
+	watchers    map[Watcher]Watcher
+	log         *logp.Logger
+}
+
+// New takes an elasticsearch client and wraps it into a fetcher, the fetch will handle the JSON
+// and response code from the cluster.
+func New(client *elasticsearch.Client, duration time.Duration, gracePeriod time.Duration) *Manager {
+	fetcher := NewElasticFetcher(client)
+	return NewWithFetcher(fetcher, duration, gracePeriod)
+}
+
+// NewWithFetcher takes a fetcher and return a license manager.
+func NewWithFetcher(fetcher Fetcher, duration time.Duration, gracePeriod time.Duration) *Manager {
```

Comment:
I think I should pass a Setting Object here so we can define the duration, grace period and anything related to the backoff.

Reply:
Are you thinking of config options that are also exposed to the user?

Reply:
I don't think we need to expose any options to the end users, they could put a really large refresh delay.. which is probably not what we want.